### PR TITLE
[PowerPC] Fix codegen for transparent_union function params

### PR DESCRIPTION
**Pull request by lei137**
Update codegen for func param with transparent_union attr to be that of the first union member.

### CODE REVIEW

**Comment by hubert-reinterpretcast** [comment]
With `--target=powerpc-unknown-linux-gnu`, the union type is still being used with `signext` (causing the IR verifier to be unhappy) for the following:
```c
typedef union __attribute__((__transparent_union__)) U {
  signed _BitInt(17) x;
} U;

__attribute__((__weak__)) long f(U u) { return u.x; }
```

I think this is primarily an artifact of `_BitInt` not being properly handled for AIX and PPC32 Linux though.

File: clang/lib/CodeGen/Targets/PPC.cpp [hunk 1]
```diff
@@ -393,6 +401,18 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
   return CharUnits::fromQuantity(4);
 }
 
+ABIArgInfo PPC32_SVR4_ABIInfo::classifyArgumentType(QualType Ty) const {
+  bool IsTransparentUnion;
+  Ty = useFirstFieldIfTransparentUnion(Ty, IsTransparentUnion);
+
+  if (IsTransparentUnion && isPromotableIntegerTypeForABI(Ty))
```
Comment (hubert-reinterpretcast) [inline clang/lib/CodeGen/Targets/PPC.cpp:408]: I don't think `isPromotableIntegerTypeForABI` handles enums. We should add a enum test case (maybe with `-fshort-enums`).

**Comment by efriedma-quic** [comment]
transparent_union generally means that the value is passed using the convention of the first member of the union.  Why do you need to check for whether the value was originally a transparent union?  Shouldn't the resulting convention be the same either way?

**Comment by lei137** [comment]
> transparent_union generally means that the value is passed using the convention of the first member of the union. Why do you need to check for whether the value was originally a transparent union? Shouldn't the resulting convention be the same either way?

That would make sense... but to extend the transparent union type based on it's first member we have call getExtend with the `llvm::type` that we want to extend to.  Otherwise it tries to extend the original union `QualType`.  The static function ` ABIArgInfo getZeroExtend()` that is called set's the ABIArgInfo to be the `llvm::type` given, otherwise it seem to default to the original if that optional param is `nullptr`.

**Comment by efriedma-quic** [comment]
Would it do any harm to just unconditionally compute the type and pass it into getExtend()?

**Comment by lei137** [comment]
> Would it do any harm to just unconditionally compute the type and pass it into getExtend()?

This seem to cause issues for type `_Bool`, where it changes the function param from `i1 noundef zeroext %b1` to `i8 noundef zeroext %b1.coerce`.  

```
$cat a.c
void fcall(_Bool);

_Bool test_wc_i1(_Bool b1, _Bool b2) {
  _Bool o=b1+b2;
  fcall(o);
  return o;
}
```
IR changes from:
```
target datalayout = "E-m:e-Fi64-i64:64-n32:64-S128-v256:256:256-v512:512:512"
target triple = "powerpc64-unknown-linux-gnu"

; Function Attrs: nounwind
define dso_local zeroext i1 @test_wc_i1(i1 noundef zeroext %b1, i1 noundef zeroext %b2) local_unnamed_addr #0 {
entry:
  %0 = or i1 %b1, %b2
  tail call void @fcall(i1 noundef zeroext %0) #3
  ret i1 %0
}

declare void @fcall(i1 noundef zeroext) local_unnamed_addr #1
```
to:
```
target datalayout = "E-m:e-Fi64-i64:64-n32:64-S128-v256:256:256-v512:512:512"
target triple = "powerpc64-unknown-linux-gnu"

; Function Attrs: nounwind
define dso_local noundef zeroext i1 @test_wc_i1(i8 noundef zeroext %b1.coerce, i8 noundef zeroext %b2.coerce) local_unnamed_addr #0 {
entry:
  %0 = or i8 %b2.coerce, %b1.coerce
  %1 = and i8 %0, 1
  %tobool = icmp ne i8 %1, 0
  tail call void @fcall(i8 noundef zeroext %1) #3
  ret i1 %tobool
}

declare void @fcall(i8 noundef zeroext) local_unnamed_addr #1
```
This is probably due to the fact I am getting the coerce type via:
```
llvm::Type *CoerceTy = llvm::IntegerType::get(getVMContext(),  getContext().getTypeSize(Ty));
```
Not sure if there is a more general way to do this?

**Comment by lei137** [comment]
I was thinking it would make the code cleaner if we can do something like this instead:
```
 // For transparent union types, return the type of the first element.
-// Set TU to true if Ty given was a transparent union and to false otherwise.
+// and set CTy the integer type of the first union element.  CTy defaults to nullptr.
-QualType CodeGen::useFirstFieldIfTransparentUnion(QualType Ty, bool &TU) {
+QualType CodeGen::useFirstFieldIfTransparentUnion(QualType Ty,
+                                                  llvm::Type *CTy) {
   if (const RecordType *UT = Ty->getAsUnionType()) {
     const RecordDecl *UD = UT->getDecl();
     if (UD->hasAttr<TransparentUnionAttr>()) {
       assert(!UD->field_empty() && "sema created an empty transparent union");
-      TU = true;
-      return UD->field_begin()->getType();
+      QualType UTy = UD->field_begin()->getType();
+      *CTy = llvm::IntegerType::get(getVMContext(),
+                                    getContext().getTypeSize(UTy));
+      return UTy;
     }
   }
-  TU = false;
   return Ty;
 }
```
Then we can just call `getExtend(Ty, Cty)` by default.  But I think I would need to put this function into `DefaultABIInfo` class to access the `getVMContext()`?  Not sure that's the right thing to do though..

**Comment by lei137** [comment]
Seems `CGT.ConvertType(Ty)` does what was needed.  I've simplified the patch to use that instead.

**Review by amy-kwan — COMMENTED** [review]
I'm not an expert here, but I think the change you made makes sense since the `Ty` is supposed to be the first field.

Also, might be a dumb question, I saw we have 32-bit Linux run lines, but is that still worth testing?

**Comment by lei137** [comment]
> I'm not an expert here, but I think the change you made makes sense since the `Ty` is supposed to be the first field.
> 
> Also, might be a dumb question, I saw we have 32-bit Linux run lines, but is that still worth testing?

Thanks for taking a look @amy-kwan.  The 32bit linux run line was added in response to review comment from Hubert: https://github.com/llvm/llvm-project/pull/101738#issuecomment-2268003106

**Review by efriedma-quic — APPROVED** [review]
LGTM

File: clang/test/CodeGen/PowerPC/transparent_union.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// RUN: %clang_cc1 -triple powerpc-unknown-linux -O2 -target-cpu pwr7 \
+// RUN:   -emit-llvm -fshort-enums %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-32
+// RUN: %clang_cc1 -triple powerpc64-unknown-aix -O2 -target-cpu pwr7 \
+// RUN:   -emit-llvm -fshort-enums %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-64
+// RUN: %clang_cc1 -triple powerpc-unknown-aix -O2 -target-cpu pwr7 \
+// RUN:   -emit-llvm -fshort-enums %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-32
+
+typedef union tu_c {
+	char a;
+	char b;
```
Comment (hubert-reinterpretcast) [inline clang/test/CodeGen/PowerPC/transparent_union.c:14]: Don't use plain `char` its (default) signedness is meant to be platform-dependent.
```suggestion
	signed char a;
	signed char b;
```

File: llvm/test/CodeGen/PowerPC/transparent_union.ll [hunk 3]
```diff
@@ -0,0 +1,123 @@
+; RUN: llc -verify-machineinstrs -mtriple=powerpc64le-unknown-linux -mcpu=pwr7 \
```
Comment (hubert-reinterpretcast) [inline llvm/test/CodeGen/PowerPC/transparent_union.ll:1]: I am not sure the machine-codegen-from-IR test is necessary. We only changed the Clang front-end IR codegen. It stands to reason that this test would have passed even with none of the other changes from this patch.

Comment (lei137) [reply llvm/test/CodeGen/PowerPC/transparent_union.ll:1]: You are right.  I verified that this pass even without this update to clang.  Will remove this IR test from the PR.

**Pull request merged**